src/tqec/circuit/detectors/boundary.py

```diff
@@ -0,0 +1,172 @@
+        self._after_collapse_cache: PauliString | None = None
+
+    @property
+    def has_anticommuting_operations(self) -> bool:
+        """Check if the instance represents a stabilizer that anti-commutes with at
+        least one of its collapsing operations.
+
+        Returns:
+            `True` if at least one collapsing operation anti-commutes with the stabilizer,
+            else `False`.
+        """
+        return bool(self._anticommuting_operations)
+
+    @property
+    def after_collapse(self) -> PauliString:
+        """Compute the stabilizer obtained after applying the collapsing operations.
+
+        Raises:
+            TQECException: If any of the collapsing operation anti-commutes with the
+                stored stabilizer.
+
+        Returns:
+            The collapsed Pauli string that goes out of the Fragment.
+        """
+        if self.has_anticommuting_operations:
+            raise TQECException(
+                "Cannot collapse a BoundaryStabilizer if it has "
+                "anticommuting operations."
+            )
+        if self._after_collapse_cache is None:
+            self._after_collapse_cache = self._stabilizer.collapse_by(
+                self._commuting_operations
+            )
+        return self._after_collapse_cache
+
+    @property
+    def before_collapse(self) -> PauliString:
+        """Return the stabilizer obtained before applying the collapsing operations.
+
+        Returns:
+            The Pauli string that goes out of the Fragment, before applying any
+            collapsing operation.
+        """
+        return self._stabilizer
+
+    @property
+    def collapsing_operations(self) -> ty.Iterator[PauliString]:
+        """Iterate on all the collapsing operations defining the boundary this
+        stabilizer is applied to.
+        """
+        yield from self._commuting_operations
+        yield from self._anticommuting_operations
+
+    @property
+    def involved_measurements(self) -> list[RelativeMeasurementLocation]:
+        return self._involved_measurements
+
+    def merge(self, other: BoundaryStabilizer) -> BoundaryStabilizer:
+        """Merge two boundary stabilizers together.
+
+        The two merged stabilizers should be defined on the same boundaries. In particular,
+        they should have the same set of collapsing operations.
+
+        Args:
+            other: the other BoundaryStabilizer to merge with self. Should have
+                exactly the same set of collapsing operations.
+
+        Returns:
+            the merged boudary stabilizer, defined on the same set of collapsing
+            operations (i.e., the same boundary), but with the two pre-collapsing
+            stabilizers multiplied together.
+        """
+        self_collapsing_operations = set(self.collapsing_operations)
+        other_collapsing_operations = set(other.collapsing_operations)
+        if self_collapsing_operations != other_collapsing_operations:
```

Comment:
The detailed logical reasoning behind that is 

Two `BoundaryStabilizer` can be merged <=> They are defined on the same boundary <=> They have the same collapsing ops

A `BoundaryStabilizer` represents a propagated stabilizer when it reaches the boundary of a fragment. Taking back the image from above, in red are the parts of the circuit where we expect boundary stabilizers.
![boundary_stab](https://github.com/QCHackers/tqec/assets/12374487/bbbf1138-b6ad-4b17-a7ff-45dd8adab040)
They basically encode, for each collapsing operation (measurement or reset), the result of propagating the stabilizer "generated" by that collapsing operation up to the opposite boundary of the fragment.

Reply:
From the order I did the review, this was not clear to me.
I understand now, but I think it is tough to understand from an outside perspective.
We should add an explicit example of how the detector computation works in the docs.

Reply:
I started to write down a section in the sphinx documentation on the different terms used in this PR, and I plan to also include in that section an example of the different steps required to automatically find detectors. This might also be as a jupyter notebook, I will see.

Reply:
The documentation is written, but will only be generated / uploaded when pushed on `main` branch. In the meantime, you can check the new files in `docs/`